process/block/poolsCleaner/miniBlocksCleaner.go

```diff
@@ -0,0 +1,153 @@
+package poolsCleaner
+
+import (
+	"sync"
+
+	"github.com/ElrondNetwork/elrond-go-logger"
+	"github.com/ElrondNetwork/elrond-go/core/check"
+	"github.com/ElrondNetwork/elrond-go/data/block"
+	"github.com/ElrondNetwork/elrond-go/dataRetriever"
+	"github.com/ElrondNetwork/elrond-go/process"
+	"github.com/ElrondNetwork/elrond-go/sharding"
+)
+
+var log = logger.GetOrCreate("process/block/poolsCleaner")
+
+const percentAllowed = 0.8
+
+// miniBlocksPoolsCleaner represents a pools cleaner that check and clean miniblocks which should not be in pool anymore
+type miniBlocksPoolsCleaner struct {
+	blockTracker     BlockTracker
+	dataPool         dataRetriever.PoolsHolder
+	rounder          process.Rounder
+	shardCoordinator sharding.Coordinator
+
+	mutMapMiniBlocksRounds sync.RWMutex
+	mapMiniBlocksRounds    map[string]int64
+}
+
+// NewMiniBlocksPoolsCleaner will return a new miniblocks pools cleaner
+func NewMiniBlocksPoolsCleaner(
+	blockTracker BlockTracker,
+	dataPool dataRetriever.PoolsHolder,
+	rounder process.Rounder,
+	shardCoordinator sharding.Coordinator,
+) (*miniBlocksPoolsCleaner, error) {
+
+	if check.IfNil(blockTracker) {
+		return nil, process.ErrNilBlockTracker
+	}
+	if check.IfNil(dataPool) {
+		return nil, process.ErrNilPoolsHolder
+	}
+	if check.IfNil(dataPool.MiniBlocks()) {
+		return nil, process.ErrNilMiniBlockPool
+	}
+	if check.IfNil(dataPool.Transactions()) {
+		return nil, process.ErrNilTransactionPool
+	}
+	if check.IfNil(rounder) {
+		return nil, process.ErrNilRounder
+	}
+	if check.IfNil(shardCoordinator) {
+		return nil, process.ErrNilShardCoordinator
+	}
+
+	mbpc := miniBlocksPoolsCleaner{
+		blockTracker:     blockTracker,
+		dataPool:         dataPool,
+		rounder:          rounder,
+		shardCoordinator: shardCoordinator,
+	}
+
+	mbpc.mapMiniBlocksRounds = make(map[string]int64)
+	miniBlocksPool := mbpc.dataPool.MiniBlocks()
+	miniBlocksPool.RegisterHandler(mbpc.receivedMiniBlock)
+
+	return &mbpc, nil
+}
+
+func (mbpc *miniBlocksPoolsCleaner) receivedMiniBlock(key []byte) {
+	if key == nil {
+		return
+	}
+
+	log.Trace("miniBlocksPoolsCleaner.receivedMiniBlock", "hash", key)
+
+	mbpc.mutMapMiniBlocksRounds.Lock()
+	defer mbpc.mutMapMiniBlocksRounds.Unlock()
+
+	if _, ok := mbpc.mapMiniBlocksRounds[string(key)]; !ok {
+		mbpc.mapMiniBlocksRounds[string(key)] = mbpc.rounder.Index()
```

Comment:
is this like a timeCache?

Reply:
Yep. A time cache combined with a pending miniblocks mechanism maintained by block tracker. So, you can consider to remove a cross miniblock from pool when both conditions are met (no cross miniblocks are in pending for me (info taken from meta) and time cache for this given cross miniblock expired). For miniblocks from me only time cache  rule is used

Reply:
ok